### PR TITLE
Upgrade to latest SPM

### PR DESCRIPTION
**Pull request by bhargavg**
Sorry for huge number of file changes, I wanted to finish this in a single PR. Majority of them are file renames & `Package.swift` updates.

> Please note that there are still few warnings related to `flatMap/compactMap` , `substring` methods. I'll try to fix them in separate PR.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,23 +5,7 @@ matrix:
...
-      services: docker
-    - script:
-        - bin/fetch-configlet
-        - bin/configlet lint .
-        - docker run -v `pwd`:/swift norionomura/swift:swift-4.1-branch bash -c "cd /swift && ./xswift-test-spm"
-      env: JOB=Linux41x
-      sudo: required
-      services: docker
-    - script:
-        - bin/fetch-configlet
```
Comment (masters3d) [inline .travis.yml]: We need these:
```
        - bin/fetch-configlet
        - bin/configlet lint .
```
They check the validity of each exercise config

**Comment by masters3d** [comment]
Could you add one more build for the Root package? This is now supported by 4.2, at some point later we could get rid of the script that checks individuals packages. 

Also, do you mind checking the `Missing exercises/proverb/.gitignore`

File: .travis.yml [hunk 2]
```diff
@@ -5,23 +5,7 @@ matrix:
...
-        - bin/configlet lint .
-        - docker run -v `pwd`:/swift norionomura/swift:swift-4.2-branch bash -c "cd /swift && swift test --verbose"
```
Comment (masters3d) [inline .travis.yml]: This one tests the root package. We would like to test all individual packages and the root packages which calls into all the subpackages.

File: circle.yml [hunk 3]
```diff
@@ -1,43 +1,22 @@
...
       - run: brew update
       - run: brew outdated swiftlint || brew upgrade swiftlint
       - run: swiftlint lint --quiet | tee lintreport.json || true # Dont crash on errors
       - run: ./xswift-test-spm
-      - run: 
+      - run:
           name: Danger Systems
           when: on_success
           command: danger || true # Dont fail build on non zero exit codes
-  build920:
```
Comment (masters3d) [inline circle.yml]: Nice! I think it is time to drop support for pre 4.2 swift

File: .travis.yml (hunk 1 above)
Comment (bhargavg) [reply .travis.yml]: We have these in the `.travis.yml` file

Full contents of [`.travis.yml`](https://github.com/bhargavg/swift/blob/chore%2Fupgrade-swiftpm-4/.travis.yml):

```YAML
language: generic

matrix:
  include:
    - script:
        - bin/fetch-configlet
        - bin/configlet lint .
        - docker run -v `pwd`:/swift swift:4.2 bash -c "cd /swift && ./xswift-test-spm"
      env: JOB=Linux42x
      sudo: required
      services: docker
```

**Comment by masters3d** [comment]
The sources are not pulled, they are referenced as of the whole repo was one package. Before this we had a Xcode project that we manually had to update. This was used mainly for development. Moving it to a single pchage allows us to create an Xcode project on the fly and updagrade swift versions in one go.

Swift overload resolution is great but avoiding duplicate global symbols is a feature in my book. 

Mainly for development.

**Comment by masters3d** [comment]
Thank you!

**Pull request merged**